demo-circuit/src/blaze_csw/constraints/mod.rs

```diff
@@ -0,0 +1,1677 @@
+                let first_half = &csw_data_g.last_wcert_g.custom_fields_g[0];
+                let second_half = csw_data_g.last_wcert_g.custom_fields_g[1]
+                    .mul_by_constant(cs.ns(|| "2^128 * custom_fields[1]"), &pow)?;
+
+                first_half.add(
+                    cs.ns(|| "custom_fields[0] + (2^128) * custom_fields[1]"),
+                    &second_half,
+                )
+            }?;
+
+            csw_data_g
+                .utxo_data_g
+                .conditionally_enforce_utxo_withdrawal(
+                    cs.ns(|| "enforce utxo withdrawal"),
+                    &scb_new_mst_root,
+                    &csw_data_g.sys_data_g.nullifier_g,
+                    &csw_data_g.sys_data_g.amount_g,
+                    &should_enforce_utxo_withdrawal_g,
+                )?;
+        }
+
+        // Enforce FT withdrawal if required
+
+        csw_data_g.ft_data_g.conditionally_enforce_ft_withdrawal(
+            cs.ns(|| "conditionally enforce ft withdrawal"),
+            &sidechain_id_g,
+            self.range_size,
+            &csw_data_g.sys_data_g.mcb_sc_txs_com_end_g,
+            &csw_data_g.sys_data_g.nullifier_g,
+            &csw_data_g.sys_data_g.amount_g,
+            &should_enforce_ft_withdrawal_g,
+        )?;
+
+        // We check the public key ownership just once for both, choosing the appropriate public key
+        // and secret key, as it is an expensive check, we want to do it just once.
+        // NOTE: We could've done the same for nullifier and amount checks, but we didn't in order
+        //       to have cleaner code (we lose only 2 constraints anyway)
+
+        ScPublicKeyGadget::enforce_pk_ownership(
+            cs.ns(|| "enforce pk ownership"),
+            &csw_data_g,
+            &should_enforce_utxo_withdrawal_g,
+        )?;
+
+        // Let's build up the public inputs
+
+        // Allocate constant as public input if needed and don't use it
+        if self.constant.is_some() {
+            let _ = FieldElementGadget::alloc_input(cs.ns(|| "alloc constant as input"), || {
+                Ok(self.constant.unwrap())
+            })?;
+        }
+
+        // Deserialize a FieldElement out of amount_g and receiver_g
+        let amount_and_receiver_fe_g = {
+            let mut amount_and_receiver_bits_g = csw_data_g
+                .sys_data_g
+                .amount_g
+                .to_bits_with_length_restriction(
+                    cs.ns(|| "amount to bits"),
+                    FieldElement::size_in_bits() - 64,
+                )?;
+
+            let mut receiver_g_bits = csw_data_g.sys_data_g.receiver_g;
+            receiver_g_bits.reverse();
+
+            amount_and_receiver_bits_g.extend_from_slice(&receiver_g_bits[..]);
+
+            FieldElementGadget::from_bits(
+                cs.ns(|| "read field element out of amount and bits"),
+                amount_and_receiver_bits_g.as_slice(),
+            )
```

Comment:
It seems to me that we implicitly assume that the length of `amount_and_receiver_bits_g` does not exceed the capacity of the constraint field. For sure this is the case in our bn_382 instantiation, but it might be worth to put an `assert!` here.